### PR TITLE
Make ParameterHandler::print_parameters() const.

### LINKED ISSUE #3989: ParameterHandler::print_parameters() and friends should be const
subj. I did not dig into the code, but I see no reason why they should not be declared as `const`.

### PR DESCRIPTION
**Pull request by bangerth**
This is done by introducing a new, private function, that duplicates most
of what the now deprecated function ParameterHandler::print_parameters_section()
does, but without changing the state of the object.

The duplication of code will be removed once the deprecated function is removed.

This fixes #3989. It passes the testsuite, which calls print_parameters()
and now the newly added function many many times.

### CODE REVIEW

File: source/base/parameter_handler.cc [hunk 1]
```diff
@@ -2311,13 +2319,391 @@ ParameterHandler::print_parameters (std::ostream     &out,
...
+        &&
+        (!(style & 128))
+        &&
+        (n_parameters != 0)
+        &&
+        (n_sections != 0))
+      out << std::endl << std::endl;
+  }
+
+  // now traverse subsections tree, in alphabetical order
```
Comment (drwells) [inline source/base/parameter_handler.cc]: Is this really in alphabetical order?

**Comment by drwells** [comment]
My criticism is mostly directed at the current implementation, not new stuff, so aside from documenting the magic numbers this is okay.

File: source/base/parameter_handler.cc [hunk 2]
```diff
@@ -2311,13 +2319,391 @@ ParameterHandler::print_parameters (std::ostream     &out,
...
+void
+ParameterHandler::recursively_print_parameters (const std::vector<std::string> &target_subsection_path,
+                                                const OutputStyle               style,
+                                                const unsigned int              indent_level,
+                                                std::ostream                   &out) const
+{
+  AssertThrow (out, ExcIO());
+
+  Assert ((style != XML) && (style != JSON),
+          ExcMessage ("This function should not be necessary for XML or JSON output."));
```
Comment (bangerth) [inline source/base/parameter_handler.cc]: This wasn't really intended as a user-facing  error message. The function in question here is `private`, and no other function in this class should call it with these `style` arguments. I'll turn this into an `ExcInternalError` and move the message into a comment.

File: source/base/parameter_handler.cc [hunk 3]
```diff
@@ -2311,13 +2319,391 @@ ParameterHandler::print_parameters (std::ostream     &out,
...
+        if (is_parameter_node (p->second) == true)
+          {
+            const std::string value = p->second.get<std::string>("value");
+
+            // if there is documentation, then add an empty line
+            // (unless this is the first entry in a subsection), print
+            // the documentation, and then the actual entry; break the
+            // documentation into readable chunks such that the whole
+            // thing is at most 78 characters wide
+            if ((!(style & 128)) &&
```
Comment (bangerth) [inline source/base/parameter_handler.cc]: I don't think we have documented this. In fact, I'd very much like to remove this obscure feature. Would you mind letting me doing this as a follow-up?

File: source/base/parameter_handler.cc [hunk 4]
```diff
@@ -2311,13 +2319,391 @@ ParameterHandler::print_parameters (std::ostream     &out,
...
+                !p->second.get<std::string>("documentation").empty())
+              {
+                if (first_entry == false)
+                  out << std::endl;
```
Comment (bangerth) [inline source/base/parameter_handler.cc]: We do. I've changed all I could find now. In the end, I'm not terribly worried -- this isn't a performance critical function. But it's worth changing anyway.

File: source/base/parameter_handler.cc [hunk 5]
```diff
@@ -2311,13 +2319,391 @@ ParameterHandler::print_parameters (std::ostream     &out,
...
+    case ShortText:
+    {
+      // first find out the longest entry name to be able to align the
+      // equal signs to do this loop over all nodes of the current
+      // tree, select the parameter nodes (and discard sub-tree nodes)
+      // and take the maximum of their lengths
+      std::size_t longest_name = 0;
+      for (boost::property_tree::ptree::const_iterator
+           p = current_section.begin();
+           p != current_section.end(); ++p)
```
Comment (bangerth) [inline source/base/parameter_handler.cc]: Hm, not trivial apparently. This doesn't seem to compile:
```
for (boost::property_tree::ptree::const_iterator p : current_section)
```
and nor does
```
for (auto p : current_section)
```
I get error messages that read:
```
/home/fac/f/bangerth/p/deal.II/1/dealii/source/base/parameter_handler.cc:2363:60: error: conversion from ‘const std::pair<const std::basic_string<char>, boost::property_tree::basic_ptree<std::basic_string<char>, std::basic_string<char> > >’ to non-scalar type ‘boost::property_tree::basic_ptree<std::basic_string<char>, std::basic_string<char> >::const_iterator’ requested
       for (boost::property_tree::ptree::const_iterator p : current_section)
```
and
```
/home/fac/f/bangerth/p/deal.II/1/dealii/source/base/parameter_handler.cc:2364:33: error: base operand of ‘->’ has non-pointer type ‘std::pair<const std::basic_string<char>, boost::property_tree::basic_ptree<std::basic_string<char>, std::basic_string<char> > >’                                                                                                
         if (is_parameter_node (p->second) == true)                                                                                                                               
                                 ^                                                                                                                                              
```
Given that some non-trivial conversion appears to be going on here, I'll leave the code as is to make the intent as clear as I can.

File: source/base/parameter_handler.cc (hunk 1 above)
Comment (bangerth) [inline source/base/parameter_handler.cc]: Yes -- we use `current_section.ordered_begin()` to get an iterator that uses alphabetical order.

**Comment by bangerth** [comment]
With the exception of the C++11-style `for` loops, I've done as you suggested, @drwells . Care to take another look?

File: source/base/parameter_handler.cc (hunk 3 above)
Comment (drwells) [inline source/base/parameter_handler.cc]: Sure.

Comment (bangerth) [inline source/base/parameter_handler.cc]: See #4500 .

File: source/base/parameter_handler.cc (hunk 5 above)
Comment (masterleinad) [inline source/base/parameter_handler.cc]: For me the following is working:

```
for (const auto p: current_section)
  if (is_parameter_node (p.second) == true)
    longest_name = std::max (longest_name, 
                             demangle(p.first).length());
```
Actually, you just have to switch `p->` to `p.` in your second try.

Comment (bangerth) [inline source/base/parameter_handler.cc]: I think I misinterpreted the C++11 for loop to say that `p` is an iterator, but it is really a reference to the elements.

Either way, I find the current code neither wrong nor difficult to read. Furthermore, the loops in the rest of the function cannot be written in this style, so I'll propose to really just leave it at that.

Comment (masterleinad) [inline source/base/parameter_handler.cc]: Sure, I just wanted to clarify that it is possible to use a range-based for loop here.

**Comment by drwells** [comment]
This looks good. Do you want to leave the `boost::io::ios_flags_saver` for later? There are a few other places in the library where we could improve our stream flag restoring, so I don't mind spinning that off as a separate issue.

**Comment by bangerth** [comment]
Oh, yes, I had forgotten about this. Let me do this in a follow-up PR.

**Comment by bangerth** [comment]
Now #4503.

**Pull request merged**